util/gctuner/tuner.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gctuner
+
+import (
+	"math"
+	"os"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/util"
+)
+
+const (
+	// MaxGCPercent is the default max cost of memory.
+	MaxGCPercent uint32 = 500
+	// MinGCPercent is the default min cost of memory.
+	MinGCPercent uint32 = 20
+)
+
+var defaultGCPercent uint32 = 100
+
+// EnableGOGCTuner is to control whether enable the GOGC tuner.
+var EnableGOGCTuner atomic.Bool
+
+func init() {
+	if val, err := strconv.Atoi(os.Getenv("GOGC")); err == nil {
+		defaultGCPercent = uint32(val)
+	}
+}
+
+// SetDefaultGOGC is to set GOGC defaultGCPercent
+func SetDefaultGOGC() {
+	util.SetGOGC(int(defaultGCPercent))
+}
+
+// Tuning sets the threshold of heap which will be respect by gc tuner.
+// When Tuning, the env GOGC will not be take effect.
+// threshold: disable tuning if threshold == 0
+func Tuning(threshold uint64) {
+	// disable gc tuner if percent is zero
+	if threshold <= 0 && globalTuner != nil {
+		globalTuner.stop()
+		globalTuner = nil
+		return
+	}
+
+	if globalTuner == nil {
```

Comment:
When will this globalTuner to be nil?

Reply:
This global tuner is nil when it startup or it is diable before. When the application starts, we must use this function to init the tuner param so that it will be inited.

Reply:
BTW, using ```func init()``` is not a good code style.